src/software/networking/udp/udp_listener.h

```diff
@@ -0,0 +1,101 @@
+#pragma once
+
+#include <boost/asio.hpp>
+
+typedef std::function<void(const char*, const size_t&)> ReceiveCallback;
+
+/**
+ * Creates a UDP listener that can listen on a given port and interface.
+ */
+class UdpListener
+{
+   public:
+    /**
+     * Creates a UDP listener.
+     *
+     * A user must check the user-provided error string to see if the listener
+     * construction was successful.
```

Comment:
Not sure if this is the best API. I think this is a case where it would be fine to throw an exception instead of forcing the user to check an error string before calling any methods on `UdpListener`. It seems like the error string is only set during construction, so wouldn't it be better just throw an exception and entirely prevent using `UdpListener` in an invalid state (i.e RAII)?

Reply:
Oh yes that might be better...

Previously we used `LOG(FATAL)` calls if the socket had an error. This would cause the program to exit. In python, this pybinded call would cause some very weird behaviour (the interpreter would get very confused and fault with undecipherable errors)

But I didn't consider a regular exception, which may be better and pybind could handle

Reply:
done